controllers/topics/getLatestPost.js

```diff
@@ -0,0 +1,52 @@
+const {QueryTypes} = require('sequelize');
+const {sequelize, Topics} = require('../../databases/models');
+
+module.exports = async (req, res) => {
+  const {name} = req.query;
+  try {
+    let topics = await Topics.findOne({
+      where: {
+        name: name
+      }
+    });
+
+    if (!topics) {
+      res.status(400).json({
+        code: 400,
+        message: 'Topic not found'
+      });
+    }
+
+    let post_topics = await sequelize.query(
+      `SELECT
+        pt.topic_id,
+        p.*
+      FROM posts p 
+        INNER JOIN post_topics pt 
+        ON p.post_id = pt.post_id
+      WHERE 
+        pt.topic_id = :topic_id
+        AND p.duration >= now()
+      ORDER BY created_at DESC
+      LIMIT 1`,
+      {
+        type: QueryTypes.SELECT,
+        replacements: {
+          topic_id: topics.topic_id
+        }
+      }
+    );
+
+    res.status(200).json({
+      status: 'success',
+      code: 200,
+      data: post_topics
+    });
+  } catch (error) {
+    console.log(error);
+    res.status(500).json({
+      code: 500,
+      message: 'Internal server error'
+    });
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

Logging the error to the console is not a good practice for production code. Instead, consider using a logging library that can provide more context about the error and handle it in a more sophisticated way. Also, it would be helpful to send some information about the error in the response to help with debugging.

Here are the suggested changes:

```diff
const {QueryTypes} = require('sequelize');
const {sequelize, Topics} = require('../../databases/models');

const getTopicByName = async (name) => {
  return await Topics.findOne({
    where: {
      name: name
    }
  });
};

const getLatestPostByTopicId = async (topicId) => {
  return await sequelize.query(
    `SELECT
      pt.topic_id,
      p.*
    FROM posts p 
      INNER JOIN post_topics pt 
      ON p.post_id = pt.post_id
    WHERE 
      pt.topic_id = :topic_id
      AND p.duration >= now()
    ORDER BY created_at DESC
    LIMIT 1`,
    {
      type: QueryTypes.SELECT,
      replacements: {
        topic_id: topicId
      }
    }
  );
};

module.exports = async (req, res) => {
  const {name} = req.query;
  try {
    let topics = await getTopicByName(name);

    if (!topics) {
      return res.status(404).json({
        code: 404,
        message: 'Topic not found'
      });
    }

    let post_topics = await getLatestPostByTopicId(topics.topic_id);

    res.status(200).json({
      status: 'success',
      code: 200,
      data: post_topics
    });
  } catch (error) {
    // Use a logging library here
    res.status(500).json({
      code: 500,
      message: `Internal server error: ${error.message}`
    });
  }
};
```
This way, the function is more maintainable and easier to test. Each function has a single responsibility which makes it easier to understand what each part of the code is doing.


<!-- This is an auto-generated comment by OSS CodeRabbit -->